app/display/model/src/main/java/org/csstudio/display/builder/model/DisplayModel.java

```diff
@@ -167,6 +171,40 @@ public final boolean isTopDisplayModel()
         return embedder == null;
     }
 
+    public final void setReaderResult(final ModelReader modelReader)
+    {
+        if (this.clean != null)
+            throw new RuntimeException("Cannot change cleanliness of DisplayModel");
+
+        this.clean = new Boolean(modelReader.getNumberOfWidgetErrors() == 0);
+    }
+
+    /** @return <code>true</code> if this display was loaded without errors,
+     *          <code>false</code> if there were widget errors
+     */
+    public final boolean isClean()
+    {
+        if (clean == null)
+            return true;
+

```

Comment:
Does 'clean' need to be volatile, then making a safe copy before checking?
```
final Boolean safe = clean;
if (safe == null)
  ..
else if (safe.booleanValue() ...
````

Reply:
Yes, you are right; `clean` needs to be volatile. But what would the safe copy achieve?

Reply:
Get the current 'volatile' once, then check its value against several options.
Otherwise it's possible that 'clean' changes during the execution of this code:
```
if (clean == null)
  ...
else if (clean.booleanValue()) ..
```
If 'clean' is set from other threads, it's possible that it was non-null for the first `if`, then set to null before the second `if`. Not sure if it indeed goes through those values, but when checking a changing value against several options, it's best to get a safe copy of the value, which can then not change while you're testing it.

Reply:
I see. In this case `clean` should not become null once it is set but I will add the safe copy to be safe

Reply:
I think you're right. `clean` is first null, then non-null, without going back to null. Still, I think it's a useful habit to keep the check 'atomic' by checking one and the same value (the safe copy) and not a potentially changing value. That'll always we predictable without having to check other setters of the value across the codebase. Thanks!